rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md

```diff
@@ -0,0 +1,342 @@
+    parts: Vec<String>,
+}
+```
+
+## Drawbacks
+
+- By increasing the time during which events are not visible on the
+  destination store, this will increase the chances of data being
+  inaccessable due to crashes and other interruptions. This is not
+  strictly data loss, as the uploaded data is recoverable by other
+  means.
+
+## Prior Art
+
+The [S3 output plugin](https://docs.fluentd.org/output/s3) for fluentd
+creates one upload per hour by default. It uses a disk buffer to store
+the object before uploading.
+
+## Alternatives
+
+### Buffer the full large object locally
+
+Instead of using the multipart upload API, Vector could save the data
+that would compose the final object into a local disk buffer, and then
+upload it in the background. This has several disadvantages:
+
+1. It requires local storage for the complete object.
+2. It creates complications for handling upload failures, as
+   acknowledgements for that upload are disconnected from any source
+   data.
+
+### Upload parts as regular objects
+
+Vector could also upload the individual batches as regular objects and
+then go back and assemble the result into a larger object after all
+the uploads are complete. Much of the proposal described above would
+be the same, but the ordering would be changed and an additional step
+is required:
+
+1. Upload batches as regular objects, making note of object names.
+2. When the final object is ready, create it with
+   `CreateMultipartUpload`.
+3. Add the previous objects using `UploadPartCopy`.
+4. Assemble the final object with `CompleteMultipartUpload`.
+5. Delete the intermediate objects that were assembled into the final
+   object.
+
+The downside with this scheme is the additional complexity of tracking
+the intermediate object names and the deletion step at the end of the
+process, both of which are handled automatically when uploading parts
+as above. Additionally, processes that monitor the destination bucket
+would end up seeing the events twice.
+
+### Use the multipart upload copy to effect append behavior
+
+One of the actions possible when creating a multipart object is to
+copy data from an existing object. This can be used to both copy data
+from and replace an object at the same time, allowing for effective
+append (or prepend) behavior without using extra bandwidth. It is
+unclear, however, if this behavior will interact badly with object
+tracking sources like SQS, or if AWS would penalize this action
+pattern for high volume uploads. As above, the data would end up being
+visible multiple times.
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is approved:
+
+- [ ] Set up the new partitioned batcher.
+- [ ] Initial MVP implementation in S3 with only the simplest upload
+      interval and no persistent buffers.
+- [ ] Add support for all upload interval types.
+
+## Future Improvements
```

Comment:
I'd like to also mention improvements to object finalization. What happens if a Vector instance crashes and never comes back? How will that object be finalized?